src/Lecture2.hs

```diff
@@ -261,7 +314,23 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval variables expression = 
+  case expression of 
+    Var variable -> 
+       case lookup variable variables of
+          Just value -> Right value
+          Nothing -> Left (VariableNotFound variable)
+
+    Lit x -> 
+      Right x
+
+    Add a b -> evalSum (eval variables a) (eval variables b)
+
+
+evalSum :: Either EvalError Int -> Either EvalError Int -> Either EvalError Int
+evalSum (Left x) _ = Left x
+evalSum _ (Left x) = Left x
+evalSum (Right x) (Right y) = Right (x + y)
```

Comment:
Once you learn about Monads, you'll be able to reduce this function to a single line 🙂